File: 0882-peak-index-in-a-mountain-array/0882-peak-index-in-a-mountain-array.cpp
class Solution {
public:
    int peakIndexInMountainArray(vector<int>& arr) {
        int n = arr.size();

        int st = 1, end  = n - 2;

        while(st <= end)
        {
            int mid = st + (end - st) / 2;
            if(arr[mid - 1] < arr[mid] && arr[mid] > arr[mid + 1]) return mid;

            if(arr[mid - 1] < arr[mid]){
                //left -> search on right side
                st = mid + 1;
            }else
            {
                //right -> search on left side
                end = mid - 1;
            }
        }
        return -1;
    }
};

File: 0882-peak-index-in-a-mountain-array/0882-peak-index-in-a-mountain-array.py
class Solution:
    def peakIndexInMountainArray(self, arr: List[int]) -> int:
        n = len(arr)
        st = 1
        end = n - 2

        while st <= end:
            mid = st + (end - st) // 2

            if arr[mid - 1] < arr[mid] and arr[mid] > arr[mid+1]:
                return mid
            
            if arr[mid - 1] < arr[mid]:
                st = mid + 1
            else:
                end = mid - 1
        
        return -1